x/pairing/keeper/scores/geo_req.go

```diff
@@ -0,0 +1,216 @@
+
+	return cost
+}
+
+func (gr GeoReq) GetName() string {
+	return geoReqName
+}
+
+// Equal() used to compare slots to determine slot groups
+func (gr GeoReq) Equal(other ScoreReq) bool {
+	otherGeoReq, ok := other.(GeoReq)
+	if !ok {
+		return false
+	}
+
+	return otherGeoReq.Geo == gr.Geo
+}
+
+func GetGeoReqsForSlots(policy planstypes.Policy) []GeoReq {
+	geoReqsForSlots := []GeoReq{}
+
+	policyGeoEnums := types.GetGeolocationsFromUint(int32(policy.GeolocationProfile))
+	switch {
+	// TODO: implement the case below
+	// case len(policyGeoEnums) > int(policy.MaxProvidersToPair):
+	default:
+		for i := 0; i < int(policy.MaxProvidersToPair); i++ {
+			geoReq := GeoReq{Geo: uint64(policyGeoEnums[i%len(policyGeoEnums)])}
+			geoReqsForSlots = append(geoReqsForSlots, geoReq)
+		}
+	}
+
+	return geoReqsForSlots
+}
+
+// a single geolocation and the latency to it (in millieseconds)
+type GeoLatency struct {
+	geo     planstypes.Geolocation
+	latency uint64
+}
+
+func (gl GeoLatency) Less(other GeoLatency) bool {
+	return gl.latency < other.latency
+}
+
+// GetGeoCost() finds the minimal latency between the required geo and the provider's supported geolocations
+func GetGeoCost(reqGeo planstypes.Geolocation, providerGeos []planstypes.Geolocation) (minLatencyGeo planstypes.Geolocation, minLatencyCost uint64) {
+	geoLatencies := []GeoLatency{}
+	latencies := []uint64{}
+	for _, pGeo := range providerGeos {
+		geoLatency := getGeoLatency(reqGeo, pGeo)
+		if geoLatency.latency == 0 {
+			continue
+		}
+		geoLatencies = append(geoLatencies, geoLatency)
+		latencies = append(latencies, geoLatency.latency)
+	}
+
+	// no geo latencies found -> provider can't support this geo (score = 1 -> max latency)
+	if len(geoLatencies) == 0 {
+		return -1, calculateCostFromLatency(maxGeoLatency)
+	}
+
+	minIndex := commontypes.FindMin(latencies)
+	minLatencyGeo = geoLatencies[minIndex].geo
+	minLatencyCost = calculateCostFromLatency(geoLatencies[minIndex].latency)
+	return minLatencyGeo, minLatencyCost
+}
+
+func getGeoLatency(from planstypes.Geolocation, to planstypes.Geolocation) GeoLatency {
+	costList := GEO_LATENCY_MAP[from]
+	for _, geoLatency := range costList {
+		if geoLatency.geo == to {
+			return geoLatency
+		}
+	}
+
+	return GeoLatency{}
+}
+
+func calculateCostFromLatency(latency uint64) uint64 {
+	return maxGeoLatency / latency
+}
+
+// define shortened names for geolocations (for convinience only)
+var (
+	USC = planstypes.Geolocation_USC
+	EU  = planstypes.Geolocation_EU
+	USE = planstypes.Geolocation_USE
+	USW = planstypes.Geolocation_USW
+	AF  = planstypes.Geolocation_AF
+	AS  = planstypes.Geolocation_AS
+	AU  = planstypes.Geolocation_AU
+)
+
+// GEO_LATENCY_MAP is a map of lists of GeoLatency that defines the cost of geo mismatch
+// for each single geolocation. The map key is a single geolocation and the value is an
+// ordered list of neighbors and their latency (ordered by latency)
+// latency data from: https://wondernetwork.com/pings
```

Comment:
mention the date when the data was taken from there (I expect it to change over time).

Reply:
done. see https://github.com/lavanet/lava/commit/b5caee08e643dacb88cea147ec151e5033a28bbc